graph/src/components/store/entity_cache.rs

```diff
@@ -209,11 +209,59 @@ impl EntityCache {
             id_is_bytes: id_is_bytes,
         };
 
+        // Get the entities from the store and insert it into the cache if it's not already there
         let entities = self.store.get_derived(&query)?;
-        entities.iter().for_each(|(key, e)| {
-            self.current.insert(key.clone(), Some(e.clone()));
+
+        entities.iter().for_each(|(k, v)| {
+            // Only insert to the cache if it's not already there
+            if !self.current.contains_key(k) {
+                self.current.insert(k.clone(), Some(v.clone()));
+            }
+        });
+
+        // Insert `None` for entities that are not in the store and not in the current cache
+        // but are present in updates or handler_updates
+        // The updates will be applied before returning the entities
+        self.updates.iter().for_each(|(k, _)| {
+            if !self.current.contains_key(k) {
+                self.current.insert(k.clone(), None);
+            }
         });
-        let entities: Vec<Entity> = entities.values().cloned().collect();
+
+        self.handler_updates.iter().for_each(|(k, _)| {
+            if !self.current.contains_key(k) {
+                self.current.insert(k.clone(), None);
+            }
+        });
+
+        // Retrieve the derived entities from the cache by filtering the cached entities.
+        // Include entities if either a matching value is found in the specified entity field
+        // as indicated by `query`, or if the entity is None.
+        let entities = self.current.iter().filter(|(key, entity)| {
+            key.entity_type == query.entity_type
+                && entity.as_ref().map_or(true, |e| {
+                    e.get(&query.entity_field.to_camel_case())
+                        .map(|v| v.to_string() == query.value.to_string())
```

Comment:
Is there any way to avoid the `to_string()` here? It seems a bit sketchy to convert these values to strings, and also causes a lot of allocations.

Reply:
How does this sound?

 ```
   .map(|v| match v {
                            Value::String(s) => s.as_str() == query.value.as_str(),
                            Value::Bytes(b) => b.as_slice() == query.value.as_bytes(),
                            _ => false,
     })
```

Reply:
Is that different from `v == query.value` ?

Reply:
yes query.value is of type `Word` and v is of type `Value`